### PR TITLE
HTTP backend user-defined headers

### LINKED ISSUE #1427: Support Authorization Bearer token for HTTP backend
### OpenTofu Version

```shell
OpenTofu v1.6.0
on linux_amd64
```

### Use Cases

`tofu` is used from CI/CD workflows where an HTTP state backend is used; the choice of backend being HTTP because not everyone can provision a cloud service based backend and could be using `tofu` for managing infrastructure that may not even be public cloud.

The current HTTP backend supports `username`  and `password` with Basic authentication. OIDC is a popular choice for credential-less authorized interactions between machines. The auth token used with OIDC is a JWT, which is generally exchanged using the `Authorization: Bearer <token>` header. Introducing this feature would allow the JWT to set as it is instead of pushing it into a Base64 structure.

### Attempted Solutions

Currently, we spoof the username field and pass the bearer token in the password field so that tofu/terraform can base64 encode it through the httpClient.  

### Proposal

- Support `bearer_token` as an optional input for the HTTP backend.
- The validation should ensure that both username + password, and `bearer_token` must not be present together.
- Build the httpClient adding the `Authorization: Bearer <bearer_token>` header, if set.

### References

_No response_

### PR DESCRIPTION
**Pull request by bunniseng**
Enables users to define headers on the HTTP backend configuration, when supplied the headers are attached to all HTTP requests to the HTTP remote backend.

Changes
- Added headers as a configuration item when creating the HTTP backend
- Added attaching user-defined headers to requests sent to the HTTP backend

Resolves #1427 

## Target Release

1.7.0

### CODE REVIEW

**Comment by bunniseng** [comment]
Hi @janosdebugs 

I have an initial implementation which allows for a user to define headers when configuring the HTTP backend, if you could review and advise that would be great, I will start on the unit tests shortly.

**Comment by bunniseng** [comment]
Hi @janosdebugs 

I have an initial implementation which allows for a user to define headers when configuring the HTTP backend, if you could do an initial review and advise that would be great.

I will start on the unit tests shortly.

**Comment by bunniseng** [comment]
Hi @janosdebugs 

This is ready for review.

Thanks

File: internal/backend/remote-state/http/backend.go [hunk 1]
```diff
@@ -220,6 +247,16 @@ func (b *Backend) configure(ctx context.Context) error {
 
 	unlockMethod := data.Get("unlock_method").(string)
 
+	var headers map[string]string
+	if dv, ok := data.GetOk("headers"); ok {
+		dh := dv.(map[string]interface{})
```
[the remark this answers is not in the document]
Comment (bunniseng) [reply internal/backend/remote-state/http/backend.go:256]: data.GetOK returns the value as an interface{} type, we need to perform an assertion to restore it back to the correct type.

File: internal/backend/remote-state/http/client.go [hunk 2]
```diff
@@ -53,8 +55,14 @@ func (c *httpClient) httpRequest(method string, url *url.URL, data *[]byte, what
 	if err != nil {
 		return nil, fmt.Errorf("Failed to make %s HTTP request: %w", what, err)
 	}
-	// Set up basic auth
-	if c.Username != "" {
+
+	// Add user-defined headers first
+	for k, v := range c.Headers {
+		req.Header.Set(strings.ReplaceAll(k, " ", ""), strings.TrimSpace(v))
```
[the remark this answers is not in the document]
Comment (bunniseng) [reply internal/backend/remote-state/http/client.go]: This is a slap forehead moment :) the scenario is already handled in the validation of not allowing white space in the header names.

File: internal/backend/remote-state/http/backend.go [hunk 3]
```diff
@@ -119,6 +120,32 @@ func New(enc encryption.StateEncryption) backend.Backend {
 				DefaultFunc: schema.EnvDefaultFunc("TF_HTTP_CLIENT_PRIVATE_KEY_PEM", ""),
 				Description: "A PEM-encoded private key, required if client_certificate_pem is specified.",
 			},
+			"headers": &schema.Schema{
+				Type:     schema.TypeMap,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+				Optional: true,
+				ValidateFunc: func(cv interface{}, ck string) ([]string, []error) {
+					rk := regexp.MustCompile("[^a-zA-Z0-9-_]")
+					rv := regexp.MustCompile("[^[:ascii:]]")
```
[the remark this answers is not in the document]
Comment (bunniseng) [reply internal/backend/remote-state/http/backend.go]: I'll need to digest this one and update it appropriately, will let you know once it's ready for another look!

[the remark this answers is not in the document]
Comment (bunniseng) [reply internal/backend/remote-state/http/backend.go]: Had a look at the raw HTTP request with wireshark when running the tests, looks like the HTTP library is not encoding characters outside the us-ascii range, I've updated the regex to use hex representation of the code-points 0-127 (us-ascii) for validation, I think my previous regex `[^[:ascii:]]` may have been the same but just to be sure lets use the updated version.

**Review by cube2222 — DISMISSED** [review]
Nice one!

**Comment by bunniseng** [comment]
@janosdebugs : fixed up imports

**Comment by cube2222** [comment]
@bunniseng You need to sign off with the DCO, see https://github.com/opentofu/opentofu/pull/1487/checks?check_run_id=23874244474 for details. You need to make sure all commits are signed-off.

**Review by cube2222 — APPROVED** [review]

**Comment by cube2222** [comment]
Thanks a lot for the contribution!

**Pull request merged**